### PR TITLE
added drawing functionality to tactical maps

### PR DESCRIPTION
**Pull request by Cthulhu80**
# About the pull request

Added a drawing feature to tactical maps, this is not done by any means and there is still plenty to do. Javascript and TGUI functionality is mostly done however there are a few bugs and the access and retrieval of the images need to be heavily optimized as the current implementation was just getting the javascript part working. Once everything is correctly implemented cic should be able to announce the tactical map with the drawings to the marines.

# Explain why it's good for the game

drawing is fun

# Testing Photographs and Procedure

drawingfunc

# Changelog

:cl:
add: added drawing functionality to tactical maps
/:cl:

### CODE REVIEW

**Comment by fira** [comment]
drawing not sync'd across maps right?

i have minor concerns regarding map export being exploitable by sending any image if it's uploaded back from tgui. it's pretty cool if it can be done at all though

**Comment by FreetoZ** [comment]
> drawing not sync'd across maps right?
> 
> i have minor concerns regarding map export being exploitable by sending any image if it's uploaded back from tgui. it's pretty cool if it can be done at all though

Toolbar selection is synced but the map itself is not, so if two users opened up the map and started drawing they would not be able to see the other's map but they would share the color selection which of course would cause problems although this is easily fixable.

**Comment by FreetoZ** [comment]
I'm going to be busy for the next week so I won't be able to work on any requested changes until I get back. Ideally, it would be great if someone who is knowledgeable in JS and React/Inferno stuff can take a thorough look before I work on optimizing the access and retrieval of the pngs. Thanks.

**Review by mullenpaul — CHANGES_REQUESTED** [review]
Some parts are looking good. I think it could be greatly simplified if, instead of rebuilding the image on update - to change global state for the lines reference. This would allow the end user to get the line data, have their browser redraw it based upon the global state, and we wouldn't have to keep resending out images.

I feel that doing it this way (assuming recaching isnt an issue) - will result in a new image being sent out via the resource transit system. This will result in a delay between the map being generated and sent to the user, they will have a blank screen for a while.

I might be wrong in my thoughts - but this approach could alleviate the concerns the maintainer team have regarding uploading and forced distribution of images. All it takes is for one person to start doing it and we have a big problem on our hands.

File: code/controllers/subsystem/minimap.dm [hunk 1]
```diff
@@ -472,15 +486,104 @@ SUBSYSTEM_DEF(minimaps)
...
+/datum/tacmap/ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
+	. = ..()
+	if(.)
+		return
+
+	var/user = ui.user
+
+	switch (action)
+		if ("clearCanvas")  // boiler plate code.
+			toolbar_updated_selection = "clear"
```
[the remark this answers is not in the document]
Comment (Cthulhu80) [reply code/controllers/subsystem/minimap.dm:525]: forgot to remove that comment, was for something else.

File: code/controllers/subsystem/minimap.dm [hunk 2]
```diff
@@ -472,15 +486,104 @@ SUBSYSTEM_DEF(minimaps)
 
 	ui = SStgui.try_update_ui(user, src, ui)
 	if(!ui)
+
+		var/icon/map_icon = getFlatIcon(map_holder.map)
+		img_ref = icon2html(map_icon, user, sourceonly = TRUE)
```
[the remark this answers is not in the document]
Comment (Cthulhu80) [reply code/controllers/subsystem/minimap.dm:491]: i'll ask around

File: tgui/packages/tgui/interfaces/CanvasLayer.js [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    this.state = {
+      selection: this.props.selection,
+    };
+
+    // needs to be of type png of jpg
+    this.img = null;
+    this.imageSrc = this.props.imageSrc;
+
+    // stores the stacked lines
+    this.lineStack = [];
```
[the remark this answers is not in the document]
Comment (Cthulhu80) [reply tgui/packages/tgui/interfaces/CanvasLayer.js:20]: Sounds like a great idea, i'll give it a go.

File: tgui/packages/tgui/interfaces/TacticalMap.tsx [hunk 4]
```diff
@@ -1,23 +1,152 @@
...
+    } else {
+      return data.toolbarUpdatedSelection;
+    }
+  };
+
   return (
-    <Window title={'Tactical Map'} theme="usmc" width={650} height={680}>
+    <Window title={'Tactical Map'} theme="usmc" width={650} height={750}>
       <Window.Content>
-        <ByondUi
```
[the remark this answers is not in the document]
Comment (Cthulhu80) [reply tgui/packages/tgui/interfaces/TacticalMap.tsx:14]: not so simple with byondui.

File: tgui/packages/tgui/interfaces/CanvasLayer.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (Cthulhu80) [reply tgui/packages/tgui/interfaces/CanvasLayer.js:20]: Ended up just going the SVG route, sorry about the delay I was busy with other stuff.

**Pull request closed without merging**